web/packages/teleport/src/ironrdp/src/lib.rs

```diff
@@ -0,0 +1,382 @@
+                    }
+                    UpdateKind::PointerDefault
+                    | UpdateKind::PointerHidden
+                    | UpdateKind::PointerPosition { .. }
+                    | UpdateKind::PointerBitmap(_) => {
+                        warn!("Pointer updates are not supported");
+                        continue;
+                    }
+                }
+            }
+
+            outputs
+        };
+
+        for output in outputs {
+            match output {
+                ActiveStageOutput::GraphicsUpdate(updated_region) => {
+                    // Apply the updated region to the canvas.
+                    let (image_location, image_data) =
+                        extract_partial_image(&self.image, updated_region);
+                    self.apply_image_to_canvas(image_data, image_location, cb_context, draw_cb)?;
+                }
+                ActiveStageOutput::ResponseFrame(frame) => {
+                    // Send the response frame back to the server.
+                    let frame = Uint8Array::from(frame.as_slice()); // todo(isaiah): this is a copy
+                    let _ = respond_cb.call1(cb_context, &frame.buffer())?;
+                }
+                ActiveStageOutput::Terminate => {
+                    return Err(JsValue::from_str("Terminate should never be returned"));
+                }
+                _ => {
+                    debug!("Unhandled ActiveStageOutput: {:?}", output);
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn apply_image_to_canvas(
+        &self,
+        image_data: Vec<u8>,
+        image_location: InclusiveRectangle,
+        cb_context: &JsValue,
+        callback: &js_sys::Function,
+    ) -> Result<(), JsValue> {
+        let top = image_location.top;
+        let left = image_location.left;
+
+        let image_data = create_image_data_from_image_and_region(&image_data, image_location)?;
+        let bitmap_frame = BitmapFrame {
+            top,
+            left,
+            image_data,
+        };
+
+        let bitmap_frame = &JsValue::from(bitmap_frame);
+
+        // TODO(isaiah): return this?
+        let _ret = callback.call1(cb_context, bitmap_frame)?;
+        Ok(())
+    }
+}
+
+pub fn extract_partial_image(
+    image: &DecodedImage,
+    region: InclusiveRectangle,
+) -> (InclusiveRectangle, Vec<u8>) {
+    // PERF: needs actual benchmark to find a better heuristic
+    if region.height() > 64 || region.width() > 512 {
+        extract_whole_rows(image, region)
+    } else {
+        extract_smallest_rectangle(image, region)
+    }
+}
+
+// Faster for low-height and smaller images
+fn extract_smallest_rectangle(
+    image: &DecodedImage,
+    region: InclusiveRectangle,
+) -> (InclusiveRectangle, Vec<u8>) {
+    let pixel_size = usize::from(image.pixel_format().bytes_per_pixel());
+
+    let image_width = usize::try_from(image.width()).unwrap();
```

Comment:
```suggestion
    let image_width = usize::from(image.width());
```

All these conversions read better with an explicit type annotation on the binding IMO:
```suggestion
    let image_width: usize = image.width().into();
```